File: cpp/test_td.cpp
#include <cstdio>
#include <cstring>
#include <string>
#include <thread>
#include <chrono>
#include <atomic>
#include <mutex>
#include <condition_variable>
#include "ThostFtdcTraderApi.h"
#include "env_loader.h"
#include "fix_locale.h"

static std::string BROKER_ID, USER_ID, PASSWORD, TD_FRONT, APP_ID, AUTH_CODE;
static std::atomic<bool> g_login_ok{false};
static std::atomic<bool> g_query_done{false};
static std::mutex g_mtx;
static std::condition_variable g_cv;

class TdSpi : public CThostFtdcTraderSpi {
public:
    explicit TdSpi(CThostFtdcTraderApi *api) : api_(api) {}

    void OnFrontConnected() override {
        printf("[TD] Connected\n");
        CThostFtdcReqAuthenticateField req{};
        strncpy(req.BrokerID, BROKER_ID.c_str(), sizeof(req.BrokerID) - 1);
        strncpy(req.UserID, USER_ID.c_str(), sizeof(req.UserID) - 1);
        strncpy(req.AppID, APP_ID.c_str(), sizeof(req.AppID) - 1);
        strncpy(req.AuthCode, AUTH_CODE.c_str(), sizeof(req.AuthCode) - 1);
        api_->ReqAuthenticate(&req, 0);
    }

    void OnFrontDisconnected(int nReason) override {
        printf("[TD] Disconnected, reason: %d\n", nReason);
    }

    void OnRspAuthenticate(CThostFtdcRspAuthenticateField *p, CThostFtdcRspInfoField *pRspInfo, int nRequestID, bool bIsLast) override {
        if (pRspInfo && pRspInfo->ErrorID != 0) {
            printf("[TD] Auth failed: %d - %s\n", pRspInfo->ErrorID, pRspInfo->ErrorMsg);
            return;
        }
        printf("[TD] Auth OK\n");
        CThostFtdcReqUserLoginField req{};
        strncpy(req.BrokerID, BROKER_ID.c_str(), sizeof(req.BrokerID) - 1);
        strncpy(req.UserID, USER_ID.c_str(), sizeof(req.UserID) - 1);
        strncpy(req.Password, PASSWORD.c_str(), sizeof(req.Password) - 1);
        api_->ReqUserLogin(&req, 0);
    }

    void OnRspUserLogin(CThostFtdcRspUserLoginField *pRspUserLogin, CThostFtdcRspInfoField *pRspInfo, int nRequestID, bool bIsLast) override {
        if (pRspInfo && pRspInfo->ErrorID != 0) {
            printf("[TD] Login failed: %d - %s\n", pRspInfo->ErrorID, pRspInfo->ErrorMsg);
            return;
        }
        printf("[TD] Login OK, TradingDay: %s, FrontID: %d, SessionID: %d\n",
               pRspUserLogin->TradingDay, pRspUserLogin->FrontID, pRspUserLogin->SessionID);
        g_login_ok = true;

        CThostFtdcSettlementInfoConfirmField req{};
        strncpy(req.BrokerID, BROKER_ID.c_str(), sizeof(req.BrokerID) - 1);
        strncpy(req.InvestorID, USER_ID.c_str(), sizeof(req.InvestorID) - 1);
        api_->ReqSettlementInfoConfirm(&req, 0);
    }

    void OnRspSettlementInfoConfirm(CThostFtdcSettlementInfoConfirmField *p, CThostFtdcRspInfoField *pRspInfo, int nRequestID, bool bIsLast) override {
        if (pRspInfo && pRspInfo->ErrorID != 0) {
            printf("[TD] Settlement confirm failed: %s\n", pRspInfo->ErrorMsg);
            return;
        }
        printf("[TD] Settlement confirmed\n");
        CThostFtdcQryTradingAccountField req{};
        strncpy(req.BrokerID, BROKER_ID.c_str(), sizeof(req.BrokerID) - 1);
        strncpy(req.InvestorID, USER_ID.c_str(), sizeof(req.InvestorID) - 1);
        api_->ReqQryTradingAccount(&req, 0);
    }

    void OnRspQryTradingAccount(CThostFtdcTradingAccountField *pTradingAccount, CThostFtdcRspInfoField *pRspInfo, int nRequestID, bool bIsLast) override {
        if (pRspInfo && pRspInfo->ErrorID != 0) {
            printf("[TD] Query account failed: %s\n", pRspInfo->ErrorMsg);
            return;
        }
        if (pTradingAccount) {
            printf("[TD] Available: %.2f, Margin: %.2f, FrozenCash: %.2f\n",
                   pTradingAccount->Available, pTradingAccount->CurrMargin, pTradingAccount->FrozenCash);
        }
        if (bIsLast) {
            std::this_thread::sleep_for(std::chrono::seconds(1));
            CThostFtdcQryInvestorPositionField req{};
            strncpy(req.BrokerID, BROKER_ID.c_str(), sizeof(req.BrokerID) - 1);
            strncpy(req.InvestorID, USER_ID.c_str(), sizeof(req.InvestorID) - 1);
            api_->ReqQryInvestorPosition(&req, 0);
        }
    }

    void OnRspQryInvestorPosition(CThostFtdcInvestorPositionField *pPos, CThostFtdcRspInfoField *pRspInfo, int nRequestID, bool bIsLast) override {
        if (pRspInfo && pRspInfo->ErrorID != 0) {
            printf("[TD] Query position failed: %s\n", pRspInfo->ErrorMsg);
            return;
        }
        if (pPos && pPos->InstrumentID[0]) {
            const char *dir = (pPos->PosiDirection == '2') ? "Long" : "Short";
            printf("[TD] Position: %s | %s | Qty: %d | PnL: %.2f\n",
                   pPos->InstrumentID, dir, pPos->Position, pPos->PositionProfit);
        }
        if (bIsLast) {
            printf("[TD] Query complete, test passed\n");
            g_query_done = true;
            g_cv.notify_all();
        }
    }

private:
    CThostFtdcTraderApi *api_;
};

int main() {
    fix_locale();
    auto env = load_env("../.env");
    BROKER_ID = env["CTP_BROKER_ID"];
    USER_ID   = env["CTP_USER_ID"];
    PASSWORD  = env["CTP_PASSWORD"];
    TD_FRONT  = env["CTP_TD_FRONT"];
    APP_ID    = env["CTP_APP_ID"];
    AUTH_CODE = env["CTP_AUTH_CODE"];

    printf("============================================================\n");
    printf("openctp TTS TD connectivity test (C++)\n");
    printf("BrokerID: %s, UserID: %s\n", BROKER_ID.c_str(), USER_ID.c_str());
    printf("TD Front: %s\n", TD_FRONT.c_str());
    printf("============================================================\n");

    auto *api = CThostFtdcTraderApi::CreateFtdcTraderApi("./td_flow/");
    TdSpi spi(api);
    api->RegisterSpi(&spi);
    api->SubscribePrivateTopic(THOST_TERT_QUICK);
    api->SubscribePublicTopic(THOST_TERT_QUICK);
    api->RegisterFront(const_cast<char *>(TD_FRONT.c_str()));
    api->Init();

    {
        std::unique_lock<std::mutex> lk(g_mtx);
        if (!g_cv.wait_for(lk, std::chrono::seconds(15), [] { return g_query_done.load(); })) {
            if (!g_login_ok) {
                printf("[TD] Timeout: login failed\n");
                api->RegisterSpi(nullptr);
                api->Release();
                return 1;
            }
        }
    }

    printf("[TD] Test completed\n");
    api->RegisterSpi(nullptr);
    api->Release();
    return 0;
}

File: py/test_td.py
"""openctp TTS 交易连接测试"""
import fix_locale  # noqa: F401  must precede openctp_ctp
import os
import time
from pathlib import Path
from dotenv import load_dotenv

load_dotenv(Path(__file__).resolve().parent.parent / ".env")

from openctp_ctp import tdapi

BROKER_ID = os.getenv("CTP_BROKER_ID")
USER_ID = os.getenv("CTP_USER_ID")
PASSWORD = os.getenv("CTP_PASSWORD")
TD_FRONT = os.getenv("CTP_TD_FRONT")
APP_ID = os.getenv("CTP_APP_ID", "")
AUTH_CODE = os.getenv("CTP_AUTH_CODE", "")


class TdSpi(tdapi.CThostFtdcTraderSpi):
    def __init__(self, api):
        super().__init__()
        self.api = api
        self.login_ok = False
        self.front_id = 0
        self.session_id = 0

    def OnFrontConnected(self):
        print("[TD] 前置连接成功")
        req = tdapi.CThostFtdcReqAuthenticateField()
        req.BrokerID = BROKER_ID
        req.UserID = USER_ID
        req.AppID = APP_ID
        req.AuthCode = AUTH_CODE
        self.api.ReqAuthenticate(req, 0)

    def OnFrontDisconnected(self, nReason):
        print(f"[TD] 前置断开连接, 原因: {nReason}")

    def OnRspAuthenticate(self, pRspAuthenticateField, pRspInfo, nRequestID, bIsLast):
        if pRspInfo and pRspInfo.ErrorID != 0:
            print(f"[TD] 认证失败: {pRspInfo.ErrorID} - {pRspInfo.ErrorMsg}")
            return
        print("[TD] 客户端认证成功")
        req = tdapi.CThostFtdcReqUserLoginField()
        req.BrokerID = BROKER_ID
        req.UserID = USER_ID
        req.Password = PASSWORD
        self.api.ReqUserLogin(req, 0)

    def OnRspUserLogin(self, pRspUserLogin, pRspInfo, nRequestID, bIsLast):
        if pRspInfo and pRspInfo.ErrorID != 0:
            print(f"[TD] 登录失败: {pRspInfo.ErrorID} - {pRspInfo.ErrorMsg}")
            return
        self.front_id = pRspUserLogin.FrontID
        self.session_id = pRspUserLogin.SessionID
        self.login_ok = True
        print(f"[TD] 登录成功")
        print(f"  交易日: {pRspUserLogin.TradingDay}")
        print(f"  FrontID: {self.front_id}, SessionID: {self.session_id}")
        print(f"  最大报单引用: {pRspUserLogin.MaxOrderRef}")

        # 确认结算单
        req = tdapi.CThostFtdcSettlementInfoConfirmField()
        req.BrokerID = BROKER_ID
        req.InvestorID = USER_ID
        self.api.ReqSettlementInfoConfirm(req, 0)

    def OnRspSettlementInfoConfirm(self, pSettlementInfoConfirm, pRspInfo, nRequestID, bIsLast):
        if pRspInfo and pRspInfo.ErrorID != 0:
            print(f"[TD] 确认结算单失败: {pRspInfo.ErrorMsg}")
            return
        print("[TD] 结算单确认成功")
        # 查询资金
        req = tdapi.CThostFtdcQryTradingAccountField()
        req.BrokerID = BROKER_ID
        req.InvestorID = USER_ID
        self.api.ReqQryTradingAccount(req, 0)

    def OnRspQryTradingAccount(self, pTradingAccount, pRspInfo, nRequestID, bIsLast):
        if pRspInfo and pRspInfo.ErrorID != 0:
            print(f"[TD] 查询资金失败: {pRspInfo.ErrorMsg}")
            return
        if pTradingAccount:
            print(f"\n[TD] === 账户资金 ===")
            print(f"  可用资金: {pTradingAccount.Available:.2f}")
            print(f"  当前保证金: {pTradingAccount.CurrMargin:.2f}")
            print(f"  冻结资金: {pTradingAccount.FrozenCash:.2f}")
            print(f"  平仓盈亏: {pTradingAccount.CloseProfit:.2f}")
            print(f"  持仓盈亏: {pTradingAccount.PositionProfit:.2f}")

        if bIsLast:
            # 查询持仓
            time.sleep(1)
            req = tdapi.CThostFtdcQryInvestorPositionField()
            req.BrokerID = BROKER_ID
            req.InvestorID = USER_ID
            self.api.ReqQryInvestorPosition(req, 0)

    def OnRspQryInvestorPosition(self, pInvestorPosition, pRspInfo, nRequestID, bIsLast):
        if pRspInfo and pRspInfo.ErrorID != 0:
            print(f"[TD] 查询持仓失败: {pRspInfo.ErrorMsg}")
            return
        if pInvestorPosition and pInvestorPosition.InstrumentID:
            direction = "多" if pInvestorPosition.PosiDirection == "2" else "空"
            print(
                f"[TD] 持仓: {pInvestorPosition.InstrumentID} | "
                f"{direction} | 数量: {pInvestorPosition.Position} | "
                f"持仓盈亏: {pInvestorPosition.PositionProfit:.2f}"
            )
        if bIsLast:
            self.query_done = True
            print("\n[TD] 查询完成，交易接口测试通过")


def run_td_test():
    print("=" * 60)
    print("openctp TTS 交易连接测试")
    print(f"BrokerID: {BROKER_ID}, UserID: {USER_ID}")
    print(f"TD Front: {TD_FRONT}")
    print("=" * 60)

    api = tdapi.CThostFtdcTraderApi.CreateFtdcTraderApi("./td_flow/")
    spi = TdSpi(api)
    api.RegisterSpi(spi)
    api.SubscribePrivateTopic(tdapi.THOST_TERT_QUICK)
    api.SubscribePublicTopic(tdapi.THOST_TERT_QUICK)
    api.RegisterFront(TD_FRONT)
    api.Init()

    time.sleep(15)
    if not spi.login_ok:
        print("[TD] 超时: 未能成功登录")
    else:
        print("\n[TD] 测试完成")
    return spi.login_ok


def test_td_connectivity():
    """pytest: 交易接口连通性测试"""
    assert run_td_test(), "TD login failed within timeout"


if __name__ == "__main__":
    import sys
    sys.exit(0 if run_td_test() else 1)
